### PR TITLE
Provide thread name (and when possible gem name) on output

### PR DESCRIPTION
**Pull request by benoittgt**
gvl-tracing handles properly handles thread name for basic usage like:
```ruby
  Thread.new do
    Thread.current.name = "thread name"
  end
```

But on more complicated examples with external libraries, it is not able to properly handles thread name.
Jean had a good idea in https://github.com/ivoanjo/gvl-tracing/issues/9 with using `Thread.list`.

This commit does this:
- Store as list of json line, events we want to monitor
- Store `Thread.list` before stopping gvl tracing
- Iterate of events stored in json file and modify entries where we can find the id of a thread
- Go a little bit further when we can and provide the gem name related to the thread

This is a very naive approach and probably not super efficient. But this is a first try.

Screenshot with the new examples file:
Screenshot 2023-04-05 at 21 23 45

Screenshot of a `gvl-tracing` on a [Rails app with some multihread](https://github.com/benoittgt/load_async_rails/blob/main/app/controllers/posts_controller.rb#L6-L11):

Screenshot 2023-04-05 at 21 45 01

Opening as a draft because code need probably some rework.

### CODE REVIEW

File: ext/gvl_tracing_native_extension/gvl_tracing.c [hunk 1]
```diff
@@ -63,7 +63,7 @@ static VALUE gc_tracepoint = Qnil;
 void Init_gvl_tracing_native_extension(void) {
   rb_global_variable(&gc_tracepoint);
 
-  VALUE gvl_tracing_module = rb_define_module("GvlTracing");
+  VALUE gvl_tracing_module = rb_define_module("GvlTracingNativeExtension");
```
Comment (casperisfine) [inline ext/gvl_tracing_native_extension/gvl_tracing.c:66]: You don't have to rename this. The module can have a mix of ruby and C methods.

If there is conflict in the name, I generally define them as private methods prefixed with an underscore, e.g. `_start`.

File: lib/gvl-tracing.rb [hunk 2]
```diff
@@ -28,6 +28,50 @@
 require_relative "gvl_tracing/version"
 
 require "gvl_tracing_native_extension"
+require "json"
 
 module GvlTracing
+  class << self
+    @@path = "/tmp/gvl-tracing.json"
```
Comment (casperisfine) [inline lib/gvl-tracing.rb:35]:
```suggestion
  @path = "/tmp/gvl-tracing.json"
  class << self
```

It's better to use an instance variable, you just have to define the default outside of `class << self`.

File: lib/gvl-tracing.rb [hunk 3]
```diff
@@ -28,6 +28,50 @@
...
+
+    def set_thread_name(thread_list)
+      output_file = File.open(@@path)
+      output = output_file.read.split("\n").map do |event|
+        parse_event = JSON.parse(event)
+        thread_id = parse_event.dig("args", "name")
+
+        next(parse_event) unless thread_id
+
+        thread = thread_list.find { |t| t.native_thread_id.to_s == thread_id.strip }
```
Comment (casperisfine) [inline lib/gvl-tracing.rb:58]: It might be faster to precompute a hash e.g. `thread_list.index_by { |t| native_thread_id.to_s }`  (minus the ActiveSupport-ism). Otherwise you end up with O(N * M) complexity.

File: lib/gvl-tracing.rb [hunk 4]
```diff
@@ -28,6 +28,60 @@
 require_relative "gvl_tracing/version"
 
 require "gvl_tracing_native_extension"
+require "json"
 
 module GvlTracing
+  @path = "/tmp/gvl-tracing.json"
+  class << self
```
Comment (casperisfine) [inline lib/gvl-tracing.rb:35]:
```suggestion
  class << self
    private :_start
    private :_stop
```

File: lib/gvl-tracing.rb [hunk 5]
```diff
@@ -28,6 +28,60 @@
...
+    def start(file)
+      @path = file
+      GvlTracing._start(@path)
```
Comment (casperisfine) [inline lib/gvl-tracing.rb:38]:
```suggestion
      _start(@path)
```

File: lib/gvl-tracing.rb [hunk 6]
```diff
@@ -28,6 +28,60 @@
...
+    end
+
+    def stop
+      thread_list = Thread.list
+
+      GvlTracing._stop
```
Comment (casperisfine) [inline lib/gvl-tracing.rb:44]:
```suggestion
      _stop
```

File: lib/gvl-tracing.rb [hunk 7]
```diff
@@ -28,6 +28,60 @@
...
+      list.each_with_object({}) do |t, acc|
+        next unless t.name
+
+        acc[t.native_thread_id] = thread_label(t)
+      end
+    end
+
+    REGEX = /lib(?!.*lib)\/([a-zA-Z-]+)/
+    def thread_label(thread)
+      lib_name = thread.to_s.match(REGEX)
```
Comment (casperisfine) [inline lib/gvl-tracing.rb:80]: If your goal is to exact a gem name or something, then `thread.backtrace_locations.last.path` might be cleaner.

File: lib/gvl-tracing.rb [hunk 8]
```diff
@@ -28,6 +28,60 @@
...
+
+        parse_event["args"]["name"] = label
+        parse_event
+      end
+      File.write(@path, output.to_json)
+    end
+
+    def aggreate_thread_list(list)
+      list.each_with_object({}) do |t, acc|
+        next unless t.name
```
Comment (casperisfine) [inline lib/gvl-tracing.rb:72]: Why skip if it doesn't have a name?

The vast majority of threads aren't named. I think a good thing would be to use the name is available, and if not fallback to parsing its backtrace.

File: ext/gvl_tracing_native_extension/gvl_tracing.c [hunk 9]
```diff
@@ -91,7 +91,7 @@ static inline void render_thread_metadata(void) {
   #endif
 
   fprintf(output_file,
-    "  {\"ph\": \"M\", \"pid\": %u, \"tid\": %u, \"name\": \"thread_name\", \"args\": {\"name\": \"%lu %s\"}},\n",
+    "  {\"ph\": \"M\", \"pid\": %u, \"tid\": %u, \"name\": \"thread_name\", \"args\": {\"name\": \"%lu %s\"}}\n",
```
Comment (benoittgt) [inline ext/gvl_tracing_native_extension/gvl_tracing.c:94]: I am wondering if we still need to provide thread name here.

File: lib/gvl-tracing.rb (hunk 7 above)
Comment (benoittgt) [reply lib/gvl-tracing.rb:80]: I look at it a little bit. But `bracktrace` is an array of strings. 
The last element looks very similar to Thread instance `to_s`
```
thread.backtrace.last
=> "/Users/benoit.tigeot/.rbenv/versions/3.2.0/lib/ruby/gems/3.2.0/gems/concurrent-ruby-1.2.2/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb:333:in `block in create_worker'"
```

File: lib/gvl-tracing.rb (hunk 8 above)
Comment (benoittgt) [reply lib/gvl-tracing.rb:72]: The thing that surprise me is that
https://github.com/ivoanjo/gvl-tracing/blob/master/examples/example4.rb#L13
Is properly handled by the C code that look for thread name. But thread list is not able to get it. But in the scenario of the Rails app (like in my initial example), it works properly with `Thread.list`.

That why I skip Thread without name here. I think your proposal is a good idea, but Thread.list needs to report properly thread name like in example 4.

File: lib/gvl-tracing.rb (hunk 7 above)
Comment (casperisfine) [reply lib/gvl-tracing.rb:84]: I was pointing at `bracktrace_locations`, not `backtrace`.

Also to avoid allocation too much useless stuff: `t.backtrace_locations(1, 1).first.path`

Comment (benoittgt) [reply lib/gvl-tracing.rb:84]: I tried an the main issue is we may do not have path.

So diff is
```diff
--- a/lib/gvl-tracing.rb
+++ b/lib/gvl-tracing.rb
@@ -81,7 +81,7 @@ module GvlTracing

     REGEX = /lib(?!.*lib)\/([a-zA-Z-]+)/
     def thread_label(thread)
-      lib_name = thread.to_s.match(REGEX)
+      lib_name = thread.backtrace_locations(1, 1).first&.path&.match(REGEX)

Comment (casperisfine) [reply lib/gvl-tracing.rb:84]: Right, could be a native method.

Comment (benoittgt) [reply lib/gvl-tracing.rb:84]: Do you think we should still change it to `backtrace_locations`? From outside I see limited gains.

File: lib/gvl-tracing.rb (hunk 8 above)
Comment (benoittgt) [reply lib/gvl-tracing.rb:76]: I think you provide the answer bellow, and it is also mentioned in [doc](https://rubyapi.org/3.2/o/thread#method-i-name).
> it may set the name to pthread and/or kernel.

**Comment by casperisfine** [comment]
> This means that the name can be completely wrong, if we hit one of these cases :)

Yeah, but it's hard to solve this. We'd need a callback on thread start to eagerly assign thread names. 

Overall, code that spawn short lived threads are rare, so maybe it's an OK limitation?

**Comment by ivoanjo** [comment]
>> This means that the name can be completely wrong, if we hit one of these cases :)
>
> Overall, code that spawn short lived threads are rare, so maybe it's an OK limitation?  

Definitely it's OK to not solve it in this PR! I brought it up since it would be nice to fix at some point -- as it's a bit of a sharp edge that can confuse users -- and maybe the discussion here would shine a light on what we could do for that one :)

> Yeah, but it's hard to solve this. We'd need a callback on thread start to eagerly assign thread names. 

I've been wondering if we could approximate it during `RUBY_INTERNAL_THREAD_EVENT_READY`/`RUBY_INTERNAL_THREAD_EVENT_RESUMED`/`RUBY_INTERNAL_THREAD_EVENT_SUSPENDED` by checking if `rb_thread_current()` is still the last one we saw for this thread. Something like this?

**Comment by casperisfine** [comment]
> Something like this?

The problem is that you can only generate the name when you hold the GVL. So you could detect that the name need to be generated again, but couldn't do it from these callbacks (except for `RESUMED`).

**Comment by ivoanjo** [comment]
> The problem is that you can only generate the name when you hold the GVL. So you could detect that the name need to be generated again, but couldn't do it from these callbacks (except for RESUMED).

Right! I guess there's two parts of the problem: 
* Detecting if it is the same thread, giving it a new "tid" as needed (or `current_thread_serial` as it's currently called)
* Getting the name for a given thread

The name could still come as a best-effort thing done later (as this PR does) -- worst case, we don't get a name.

**Comment by benoittgt** [comment]
> Thus, I think we don't need to do the reparsing; instead, as part of stop, we could re-print the names of the threads to the trace file, e.g. by perhaps building up the thread names in Ruby as you're doing in the PR, and then passing it as a string that gets appended to the trace file before it gets closed. And for perfetto it seems like this is OK -- we don't even need to omit the duplicate names at the beginning.

That's interesting. I'm gonna have a look at what is possible with metadatas and perfetto UI.

**Comment by benoittgt** [comment]
@ivoanjo I had a quick look of your proposal and pushed a commit. It is much better. Good catch with perfetto trace format. The only issue I have is the `thread_id` provided by the C extension is different from the one I get in thread list. I didn't look a it yet.

File: lib/gvl-tracing.rb [hunk 10]
```diff
@@ -28,6 +28,53 @@
 require_relative "gvl_tracing/version"
 
 require "gvl_tracing_native_extension"
+require "json"
 
 module GvlTracing
+  @path = "/tmp/gvl-tracing.json"
```
Comment (ivoanjo) [inline lib/gvl-tracing.rb:34]: Minor: I think this is no longer needed?

File: lib/gvl-tracing.rb [hunk 11]
```diff
@@ -28,6 +28,53 @@
...
+
+  class << self
+    private :_start
+    private :_stop
+
+    def start(file)
+      @path = file
+      _start(@path)
+    end
```
Comment (ivoanjo) [inline lib/gvl-tracing.rb:43]: There's a bit of validation logic in `_start` which may actually make it not start, thus the `@path` should only be updated after `_start` returns without an exception:

```suggestion
    def start(file)
      _start(file)
      @path = file
    end
```

In particular, the following issue could happen before:

```ruby
GvlTracing.start("foo.json")
GvlTracing.start("bar.json") rescue nil # This fails, but the `@path` gets changed
GvlTracing.stop # The events are in `foo.json`, the thread names end up in `bar.json`
```

File: lib/gvl-tracing.rb [hunk 12]
```diff
@@ -28,6 +28,53 @@
...
+        f.puts(threads_name)
+        f.puts("]")
+      end
+    end
+
+    def aggreate_thread_list(list)
+      list.each_with_object([]) do |t, acc|
+        next unless t.name
+
+        acc << {"ph": "M", "pid": Process.pid, "tid": t.native_thread_id, "name": "thread_name", "args": {"name": thread_label(t)}}.to_json
```
Comment (ivoanjo) [inline lib/gvl-tracing.rb:67]: Minor: To be honest, this line is so close to the final output JSON, that I'm not sure it's worth using `to_json` ;)

File: lib/gvl-tracing.rb (hunk 8 above)
Comment (ivoanjo) [reply lib/gvl-tracing.rb:65]: As a note, example4.rb doesn't work with the new code because the threads die before `.stop` -- thus they won't be in `Thread.list` by the time we call it.

**Review by ivoanjo — COMMENTED** [review]
Ah, yes, the sequential `tid`! On hindsight, I did totally gloss over that in my suggestion, making it seem easier than it was -- sorry!

You can actually see the story of its introduction in https://github.com/ivoanjo/gvl-tracing/pull/4 .
TL;DR:

* `gettid()` is not supported on macOS
* We discussed replacing it by `rb_nativethread_self()` but that got us long ids which the perfetto UI didn't like at the time
* The sequential id ended up being adopted instead

It was only in a later change that I introduced the `pthread_threadid_np` for macOS as the tid.

So yes, I think at this point it's perfectly reasonable to switch back to using the `thread_id` as the `tid`, instead of the sequential value. 

(At some point something like the sequential value may need to make a return to allow distinguishing thread reuse, but let's put that aside for now.)

**Comment by benoittgt** [comment]
> So yes, I think at this point it's perfectly reasonable to switch back to using the thread_id as the tid, instead of the sequential value.

I am wondering if my change properly address your suggestion. 

I tested the last code change with success with `example4.rb` (we fallback on thread name via `native_thread_name_buffer`), `thread_name_with_extension.rb` and [Rails app code](https://github.com/benoittgt/load_async_rails/blob/91e421414e2bd3ae0687f58ddf97bb14427b44dd/app/controllers/posts_controller.rb#L6-L16).

File: ext/gvl_tracing_native_extension/gvl_tracing.c [hunk 13]
```diff
@@ -158,6 +150,20 @@ static double timestamp_microseconds(void) {
   return (current_monotonic.tv_nsec / 1000.0) + (current_monotonic.tv_sec * 1000.0 * 1000.0);
 }
 
+static uint64_t native_thread_id() {
+  uint64_t native_thread_id = 0;
+
+  #ifdef HAVE_PTHREAD_THREADID_NP
```
Comment (benoittgt) [inline ext/gvl_tracing_native_extension/gvl_tracing.c:156]: We first look at `HAVE_PTHREAD_THREADID_NP` instead of `HAVE_GETTID`.

File: ext/gvl_tracing_native_extension/gvl_tracing.c [hunk 14]
```diff
@@ -169,7 +175,7 @@ static void render_event(const char *event_name) {
     render_thread_metadata();
   }
 
-  unsigned int thread_id = current_thread_serial;
+  uint64_t thread_id = native_thread_id();
```
Comment (ivoanjo) [inline ext/gvl_tracing_native_extension/gvl_tracing.c:178]: Rather that retrieving it on every call, I suggest storing the `native_thread_id` in a `_Thread_local` like we're doing for the serial, and then initializing it only once (e.g. similar to what we're doing with the serial).

This way we avoid that potential overhead for each event emission :)

File: ext/gvl_tracing_native_extension/gvl_tracing.c [hunk 15]
```diff
@@ -143,7 +135,7 @@ static VALUE tracing_stop(VALUE _self) {
   gc_tracepoint = Qnil;
 
   render_event("stopped_tracing");
-  fprintf(output_file, "]\n");
+  // closing output file is handled in GvlTracing.stop code
```
Comment (ivoanjo) [inline ext/gvl_tracing_native_extension/gvl_tracing.c:138]: Minor: I suggest clarifying that this comment is referencing the json structure, as otherwise it can be a bit confusing without the context of this diff because the next line is an `fclose(...)` and we're commenting that we're not closing stuff ;)

```suggestion
  // closing the json syntax in the output file is handled in GvlTracing.stop code
```

File: ext/gvl_tracing_native_extension/gvl_tracing.c [hunk 16]
```diff
@@ -75,25 +76,16 @@ static inline void initialize_thread_id(void) {
...
     pthread_getname_np(pthread_self(), native_thread_name_buffer, sizeof(native_thread_name_buffer));
   #endif
 
   fprintf(output_file,
-    "  {\"ph\": \"M\", \"pid\": %u, \"tid\": %u, \"name\": \"thread_name\", \"args\": {\"name\": \"%lu %s\"}},\n",
-    process_id, current_thread_serial, native_thread_id, native_thread_name_buffer
-  );
+    "  {\"ph\": \"M\", \"pid\": %u, \"tid\": %llu, \"name\": \"thread_name\", \"args\": {\"name\": \"%llu %s\"}},\n",
+    process_id, thread_id, thread_id, native_thread_name_buffer);
 }
```
Comment (ivoanjo) [inline ext/gvl_tracing_native_extension/gvl_tracing.c:89]: The change to use `%llu` is actually not correct for Linux -- if you look at the CI runs, it complains that it's not correct of a uint64_t. (But it is correct for macOS -- going back to `%lu` makes the macOS build complain).

[I suggest using `PRIu64`](https://stackoverflow.com/questions/16859500/what-is-priu64-in-c), which should work correctly for both.

File: ext/gvl_tracing_native_extension/gvl_tracing.c [hunk 17]
```diff
@@ -75,25 +76,16 @@ static inline void initialize_thread_id(void) {
...
 
 static inline void render_thread_metadata(void) {
```
Comment (ivoanjo) [inline ext/gvl_tracing_native_extension/gvl_tracing.c:78]: Looking at the perfetto rendering for the `thread_name_with_extension.rb` example...

image

it always prints the `tid` anyway, so it gets duplicated now. Previously this was the thread serial which was different :)

So I'd suggest a slight tweak here -- printing only the name.

File: lib/gvl-tracing.rb (hunk 8 above)
Comment (ivoanjo) [reply lib/gvl-tracing.rb:62]: Maybe when `t == Thread.main` it's worth printing "Main Thread" or something?

File: lib/gvl-tracing.rb (hunk 7 above)
Comment (ivoanjo) [reply lib/gvl-tracing.rb:70]: I don't have a strong opinion on this one (yet? xD), so let's go with what you picked for now :)

File: examples/thread_name_with_extension.json [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+  {"ph": "E", "pid": 44888, "tid": 476842, "ts": 69.000000},
+  {"ph": "B", "pid": 44888, "tid": 476842, "ts": 69.000000, "name": "wants_gvl"},
+  {"ph": "M", "pid": 44888, "tid": 476843, "name": "thread_name", "args": {"name": "476843 "}},
+  {"ph": "E", "pid": 44888, "tid": 476843, "ts": 82.000000},
+  {"ph": "B", "pid": 44888, "tid": 476843, "ts": 82.000000, "name": "started"},
+  {"ph": "E", "pid": 44888, "tid": 476843, "ts": 85.000000},
+  {"ph": "B", "pid": 44888, "tid": 476843, "ts": 85.000000, "name": "wants_gvl"},
+  {"ph": "M", "pid": 44888, "tid": 476844, "name": "thread_name", "args": {"name": "476844 "}},
+  {"ph": "E", "pid": 44888, "tid": 476844, "ts": 93.000000},
+  {"ph": "B", "pid": 44888, "tid": 476844, "ts": 93.000000, "name": "started"},
```
Comment (ivoanjo) [inline examples/thread_name_with_extension.json:17]: Not anything this PR touches or changes, but it's interesting that we're not getting beyond microsecond precision on macOS, but we do so on Linux. At some point I need to perhaps look into the other options beyond `CLOCK_MONOTONIC` to see if we can make it tighter (I think there's a `_RAW` variant....)

File: ext/gvl_tracing_native_extension/gvl_tracing.c (hunk 16 above)
Comment (benoittgt) [reply ext/gvl_tracing_native_extension/gvl_tracing.c:89]: I switched to `long int`. Maybe it's better?

File: lib/gvl-tracing.rb (hunk 8 above)
Comment (benoittgt) [reply lib/gvl-tracing.rb:62]: Good idea

**Comment by benoittgt** [comment]
Thanks @ivoanjo for your review. Here is few screenshots of the new output. 

I am not sure for the C changes. You will tell me. :)

Screenshot 2023-04-27 at 17 48 29
Screenshot 2023-04-27 at 17 48 24

File: examples/thread_name_with_extension.json (hunk 18 above)
Comment (benoittgt) [reply examples/thread_name_with_extension.json]: Interesting.

**Comment by ivoanjo** [comment]
This looks great! I hope you don't take my long delay the wrong way -- I definitely I am very grateful for the contribution. It's just been an intense few weeks :)

And feel free to send more things this way :)

**Comment by ivoanjo** [comment]
@benoittgt I've just released this improvement as version 1.2.0 on rubygems :tada:
